onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -1865,6 +1865,14 @@ std::unique_ptr<IDataTransfer> TensorrtExecutionProvider::GetDataTransfer() cons
 }
 
 Status TensorrtExecutionProvider::OnRunStart(const onnxruntime::RunOptions& /*run_options*/) {
+  // The GPU device is set again here to handle multithreading scenarios.
+  // Consider the following:
+  // Users can create multiple threads to initialize separate inference sessions on different devices (not just the default device 0)
+  // Later, additional threads may be spawned to execute inference_session.Run(), which calls this compute function.
+  // Since new threads default to using device 0, it’s necessary to explicitly set the correct device to ensure computations run on the intended GPU.
+  // Note: Based on our measurements on the A100 GPU with CUDA 12, the execution time for cudaSetDevice is approximately 0.004 ms, which is negligible
+  //       and does not impact runtime performance.
+  cudaSetDevice(device_id_);
```

Comment:
Shall device_id stored in thread state instead of sharing by all threads?

Reply:
A new thread can be created and then directly run inference.Run(), in this case, i think we can only get the device id from the provider to know that GPU device to run on.

Reply:
CUDA EP also get the device_id from provider? correct me if i'm wrong.
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/cuda/cuda_execution_provider.cc#L424

Reply:
CUDA EP does not support such scenario (only allows one GPU per session).